test/exe/Main.hs

```diff
@@ -230,7 +230,8 @@ diagnosticTests = testGroup "diagnostics"
       _ <- openDoc' "ModuleB.hs" "haskell" contentB
       expectDiagnostics [("ModuleB.hs", [(DsError, (1, 7), "Could not find module")])]
       let contentA = T.unlines [ "module ModuleA where" ]
-      _ <- openDoc' "ModuleA.hs" "haskell" contentA
+      TextDocumentIdentifier newModule <- openDoc' "ModuleA.hs" "haskell" contentA
```

Comment:
This looks like it is no longer sufficient for the client to just send `openDoc'` if it supports `WorkspaceDidChangeWatchedFiles`. I don’t think that is correct. `WorkspaceDidChangeWatchedFiles` is for files on disk that aren’t opened. So the test for this should only write a file to disk and send the event, not call `openDoc'`. On the other hand, `openDoc'` should still be sufficient for this error to go away even without the event since the file might not be written to disk.

Reply:
I agree that `openDoc'` is not needed for the test, but it's still handy to construct the uri. 

I'm not sure what your second statement means. Can you elaborate? The file does not exist on disk initially, so `GetLocatedModules` will not consider after creation it until the event has been received

Reply:
Given that `openDoc'` affects the VFS which in turn affects the file exists check, I really think we should not use this here or ideally have two tests. One using `openDoc'` and one without.

As for the second question, without your PR a file is considered existent if it is in the VFS and only if it isn’t in the VFS we look it up on disk. So `openDoc` which adds it to the VFS should be sufficient for it to be picked up. With your change that seems to no longer hold.

Reply:
Iiuc the problem is that `openDoc'` is badly named. It not only "opens" the file, but it also creates it. Therefore it really should be sending watched file creation notifications - if it was, then it would be sufficient. 

Let me know if that makes sense and I can add it to the PR.

Reply:
Sorry I think I’m not making myself clear. Let me try to give a concrete example.

1. You have module `A` on disk and opened in your LSP client (let’s say Emacs since VSCode behaves in a specific way that makes this a non-issue afaik but that’s definitely not mandated by the LSP spec). Module `A` imports module `B` which is neither opened nor does it exist on dis.
2. Now you open a new buffer for module `B`. This will send a `DidOpenTextDocument` notification to the server (i.e. ghcide). This will result in the module being added to the VFS. However, the module does not necessarily exist on disk at this point. (This is where it matters that we use Emacs and not VSCode since opening a buffer for a non-existent file does not write it to disk until you save whereas I don’t know how to open a buffer for a non-existent file in VSCode since the new file function seems to create it on disk).
3. The client will not send a `DidChangeWatchedFilesParams` notification since the file does not yet exist on disk.
4. Without your PR we consider a file existent if it exists in the VFS so at this point `B` exists and the import in `A` can be resolved.
5. With your PR you will not consider `B` existent as far as I can tell.

Reply:
Thanks for the detailed explanation, I see what you mean now. Indeed, until B is saved to disk, it will not be considered as existing; this is a behaviour change, albeit only noticeable in some editors. But then again, the user only needs to save the file to make things work.

The issue is subtle - care needs to be taken in order to ensure that every new buffer is covered by a `DidChangeWatchedFile` subscription, otherwise we will miss renames/deletions. 

Since the behaviour change is minimal and the solution is tricky, I would vote for waiting until we have a user opened issue before trying to fix.

Reply:
Sounds reasonable. Let’s please add a comment (probably to the `GetFileExists` rule) that mentions this. Maybe just copy the whole example that I posted above.

Reply:
Done!